ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -0,0 +1,279 @@
+    * Produces the log entry to be committed, and DAML state updates.
+    *
+    * The caller is expected to resolve the inputs declared in [[DamlSubmission]] prior
+    * to calling this method, e.g. by reading [[DamlSubmission.getInputEntriesList]] and
+    * [[DamlSubmission.getInputStateList]]
+    *
+    * The caller is expected to store the produced [[DamlLogEntry]] in key-value store at a location
+    * that can be accessed through `entryId`. The DAML state updates may create new entries or update
+    * existing entries in the key-value store. The concrete key for DAML state entry is obtained by applying
+    * [[packDamlStateKey]] to [[DamlStateKey]].
+    *
+    * @param engine: The DAML Engine. This instance should be persistent as it caches package compilation.
+    * @param config: The ledger configuration.
+    * @param entryId: The log entry id to which this submission is committed.
+    * @param recordTime: The record time for the log entry.
+    * @param submission: The submission to commit to the ledger.
+    * @param inputLogEntries: The resolved inputs to the submission.
+    * @param inputState: The input DAML state entries. Some state entries declared as input can be missing.
+    * @return The log entry to be committed and the DAML state updates.
+    */
+  def processSubmission(
+      engine: Engine,
+      config: Configuration,
+      entryId: DamlLogEntryId,
+      recordTime: Timestamp,
+      submission: DamlSubmission,
+      inputLogEntries: Map[DamlLogEntryId, DamlLogEntry],
+      inputState: Map[DamlStateKey, Option[DamlStateValue]]
+  ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) = {
+
+    submission.getPayloadCase match {
+      case DamlSubmission.PayloadCase.ARCHIVE =>
+        val archive = submission.getArchive
+        val key = DamlStateKey.newBuilder.setPackageId(archive.getHash).build
+        (
+          DamlLogEntry.newBuilder
+            .setRecordTime(buildTimestamp(recordTime))
+            .setArchive(archive)
+            .build,
+          Map(
+            key -> DamlStateValue.newBuilder.setArchiveEntry(entryId).build
+          )
+        )
+      case DamlSubmission.PayloadCase.CONFIGURATION_ENTRY =>
+        (
+          DamlLogEntry.newBuilder
+            .setRecordTime(buildTimestamp(recordTime))
+            .setConfigurationEntry(submission.getConfigurationEntry)
+            .build,
+          Map.empty
+        )
+      case DamlSubmission.PayloadCase.TRANSACTION_ENTRY =>
+        processTransactionSubmission(
+          engine,
+          config,
+          entryId,
+          recordTime,
+          submission.getTransactionEntry,
+          inputLogEntries,
+          inputState
+        )
+
+      case DamlSubmission.PayloadCase.PAYLOAD_NOT_SET =>
+        throw new RuntimeException("DamlSubmission payload not set!")
+    }
+  }
+
+  private def processTransactionSubmission(
```

Comment:
comment

Reply:
this is a completely internal method and would just duplicate the documentation from the above public method.